Import image from dma_buf following VK_EXT_external_memory_dma_buf

DavidR86 opened the pull request:
1. [x] Update documentation to reflect any user-facing changes - in this repository.

2. [ ] Make sure that the changes are covered by unit-tests.

3. [x] Run `cargo fmt` on the changes.

4. [x] Please put changelog entries **in the description of this Pull Request**
   if knowledge of this change could be valuable to users. No need to put the
   entries to the changelog directly, they will be transferred to the changelog
   file by maintainers right after the Pull Request merge.
   
   Please remove any items from the template below that are not applicable.

5. [x] Describe in common words what is the purpose of this change, related
   Github Issues, and highlight important implementation aspects:

Implements importing an image into Vulkan from a Linux dma_buf,
according to the following Vulkan extensions:
- https://registry.khronos.org/vulkan/specs/1.3-extensions/man/html/VK_EXT_external_memory_dma_buf.html
- https://registry.khronos.org/vulkan/specs/1.3-extensions/man/html/VK_EXT_image_drm_format_modifier.html

This functionality is already part of Ash and the Vulkan specification, so not many changes were necessary. 
The main change was a new constructor to `StorageImage` that creates an image from a list of Unix file descriptors, a DRM_FORMAT_MODIFIER and some other parameters. This only adds support for importing single-planar images (those backed by a single file descriptor), despite the specification supporting multiple file descriptors.

Changelog:
```markdown
### Public dependency updates
None
 
### Breaking changes
None

### Additions
- Implements importing an image into Vulkan from a Linux dma_buf, following the `VK_EXT_external_memory_dma_buf` extension.

### Bugs fixed
None
````

Rua commented:
Can you add the appropriate validation for any VUIDs that rely on the tiling of the image? Both for `VkImageCreateInfo` and elsewhere.

DavidR86 commented:
> Can you add the appropriate validation for any VUIDs that rely on the tiling of the image? Both for `VkImageCreateInfo` and elsewhere.
> 
> The Windows build is also currently failing, that should be fixed.

I noticed that on the VUIDs validation on `ImageCreateInfo`, some checks are done with `assert!`s, and some are done by returning `ImageCreateError`s. Which option should I prefer for this?

Rua commented:
The Windows CI stuff has been intermittent for a while. This isn't caused by your code, so you don't need to worry about it.

There are many other VUIDs throughout the Vulkan specification that reference the tiling of an image. Some examples:
- VUID-vkGetPhysicalDeviceImageFormatProperties-tiling-02248
- VUID-VkPhysicalDeviceImageFormatInfo2-tiling-02249
- VUID-vkGetImageSubresourceLayout-02270

You can get a better idea by searching for `VK_IMAGE_TILING_` or `VK_EXT_image_drm_format_modifier` within the source code of the documentation, which is available at https://github.com/KhronosGroup/Vulkan-Docs.

Rua commented:
Actually, looking at the Vulkan spec closer, it seems that VUID-vkGetPhysicalDeviceImageFormatProperties-tiling-02248 is related to VUID-VkPhysicalDeviceImageFormatInfo2-tiling-02249. Since the non-2 version of the function has no way to provide `pNext` structures, it follows that VUID-VkPhysicalDeviceImageFormatInfo2-tiling-02249 can never be fulfilled when calling that function. So there may be no need to validate VUID-vkGetPhysicalDeviceImageFormatProperties-tiling-02248 separately.

DavidR86 commented:
I created a list of all VUIDs by searching for VK_IMAGE_TILING_ or VK_EXT_image_drm_format_modifier. 
I added all the ones I could (the ones that have a checkbox), but many are for vulkan calls that vulkano does not yet do anywhere in it's source code. Others also pertain muliplanar images imported using a Linux FD, which is something this PR does not currently implement. I have added a check with a panic! to the method I made, which can be removed when that is implemented.

- [x] VUID-vkGetPhysicalDeviceImageFormatProperties-tiling-02248
- [x] VUID-VkPhysicalDeviceImageFormatInfo2-tiling-02249
- [ ] VUID-VkPhysicalDeviceImageFormatInfo2-tiling-02313 (not possible)
- [x] VUID-VkImageCreateInfo-tiling-02261
- [x] VUID-VkImageCreateInfo-pNext-02262
- [ ] VUID-VkImageCreateInfo-tiling-02353 (not possible)
- [ ] VUID-vkGetImageDrmFormatModifierPropertiesEXT-image-02272 (not possible)
- [ ] VUID-VkImageMemoryRequirementsInfo2-image-02279
	- Vulkano currently has no validation in `get_memory_requirements()`  
	- No `VkImagePlaneMemoryRequirementsInfo` in `get_memory_requirements`
	- Therefore setting `VK_IMAGE_CREATE_DISJOINT_BIT` would not work, but multiplanar image importing is not supported anyways.
- [ ] VUID-VkImageMemoryRequirementsInfo2-image-02280
	- Same as VUID-VkImageMemoryRequirementsInfo2-image-02279
- [ ] VUID-VkDeviceImageMemoryRequirementsKHR-pCreateInfo-06420
	- VkDeviceImageMemoryRequirements not yet used in vulkano
- [ ] VUID-VkImagePlaneMemoryRequirementsInfo-planeAspect-02282
	- Multiplanar/disjoint
- [x] VUID-VkBindImagePlaneMemoryInfo-planeAspect-02284
	- Muliplanar/disjoint
- [ ] VUID-VkImageSubresourceLayers-aspectMask-02247 (memory planes)
- [ ] VUID-VkClearAttachment-aspectMask-02246 (memory planes)
- [ ] VUID-vkGetImageSubresourceLayout-tiling-02271
	- Hard to do rn with only single-plane images
- [ ] VUID-vkGetImageSubresourceLayout2EXT-tiling-02271
	- Same as previous
- [ ] VUID-VkInputAttachmentAspectReference-aspectMask-02250
	- Only used once, without any other checks?
- [ ] VUID-VkImageCreateInfo-pNext-06746
	- No `VkImageCompressionControlEXT` in vulkano yet.
- [ ] VUID-VkImageDrmFormatModifierListCreateInfoEXT-pDrmFormatModifiers-02263
	- `VkImageDrmFormatModifierListCreateInfoEXT` is not used anywhere, only `VkImageDrmFormatModifierExplicitCreateInfoEXT`
- [ ] VUID-VkImageDrmFormatModifierExplicitCreateInfoEXT-drmFormatModifier-02264
	- - Maybe we can check this?
- [ ] VUID-VkImageDrmFormatModifierExplicitCreateInfoEXT-drmFormatModifierPlaneCount-02265
	- Maybe we can check this?
- [x] VUID-VkImageDrmFormatModifierExplicitCreateInfoEXT-size-02267
- [x] VUID-VkImageDrmFormatModifierExplicitCreateInfoEXT-arrayPitch-02268
- [x] VUID-VkImageDrmFormatModifierExplicitCreateInfoEXT-depthPitch-02269
- [x] VUID-vkGetImageSubresourceLayout-image-07790
	- Done, same as `VUID-vkGetImageSubresourceLayout-image-02270`
- [ ] VUID-VkImageMemoryRequirementsInfo2-image-01590
	- No multiplanar support at the moment
- [ ] VUID-VkImageMemoryRequirementsInfo2-image-01591
- [ ] VUID-VkDeviceImageMemoryRequirements-pCreateInfo-06776
- [ ] VUID-VkSubpassDescription2-attachment-04563
	- multiple planes

Rua commented on vulkano/src/image/storage.rs:
```diff
@@ -285,6 +436,20 @@ impl StorageImage {
+#[cfg(target_os = "linux")]
+/// Struct that contains a Linux file descriptor for importing, when creating an image. Since a file descriptor is used for each
+/// plane in the case of multiplanar images, each fd needs to have an offset and a row pitch in order to interpret the imported data.
+pub struct SubresourceData {
```
The comments here should start with /// so that they show up as documentation. Also "hanfle".

DavidR86 commented:
Is there anything else that needs to be changed or added?

Rua commented:
You mention that there is no multiplanar support yet, but Vulkano does have support for multiplanar images. Do you mean multiplanar images specifically in combination with the extensions you're adding?
